src/java.base/share/classes/java/net/URLEncoder.java

```diff
@@ -222,20 +227,30 @@ public static String encode(String s, String enc)
     public static String encode(String s, Charset charset) {
         Objects.requireNonNull(charset, "charset");
 
-        boolean needToChange = false;
-        StringBuilder out = new StringBuilder(s.length());
+        int i;
+        for (i = 0; i < s.length(); i++) {
+            char c = s.charAt(i);
+            if (!DONT_NEED_ENCODING.test(c) || c == ' ') {
+                break;
+            }
+        }
+        if (i == s.length()) {
+            return s;
+        }
+
+        StringBuilder out = new StringBuilder(s.length() << 1);
         CharArrayWriter charArrayWriter = new CharArrayWriter();
+        if (i > 0) {
+            out.append(s, 0, i);
+        }
 
-        for (int i = 0; i < s.length();) {
-            int c = s.charAt(i);
-            //System.out.println("Examining character: " + c);
+        while (i < s.length()) {
+            char c = s.charAt(i);
             if (DONT_NEED_ENCODING.test(c)) {
                 if (c == ' ') {
                     c = '+';
```

Comment:
The extra test (on every regular character) for space could be moved to a separate if at line 255 (and remove space from DONT_NEED_ENCODING). The performance improvement might not be noticable but it would remove an anomaly from the algorithm.